### ISSUE TITLE
Match involving unpacked tuple incorrectly marked unreachable

### ISSUE THREAD

**Issue by JelleZijlstra** [python/mypy#17095]
```python
def f(x: tuple[int, str, *tuple[range, ...]]) -> None:
    match x:
        case [1, *ts]:
            reveal_type(ts)
```

([Playground](https://mypy-play.net/?mypy=latest&python=3.12&flags=warn-unreachable&gist=6b5dd7f552779d3d72b102bdc42d775b))

Produces no output, and with `--warn-unreachable` on says that line 4 is unreachable. But it's not unreachable; the case can match.

For comparison, pyright does not think the clause is unreachable but deduces an incorrect type (microsoft/pyright#7613).

**Comment by randolf-scholz** [comment]
Here's another instance I noticed:

https://mypy-play.net/?mypy=latest&python=3.11&flags=warn-unreachable&gist=415990466cbd7bf9c78c6d7f71bebfd0

[pyright-playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMpgBuApiADZgCGAJgDRQCCKc9AysQI4CuxKAxsQBQg6sWBRgYMAApKADyQBnAFyNmASmWCoOqBEow%2BACyjylW3Zah9Ki4lAByYFMQtX3IYjC4gUj50LuOjZ2UADaALpuQbqe3r5Q0uo6AMRQgDLkUACqKJ6UxpQARmSBQSH2qDCymtoxsV4%2BfrK06rXu5VAA%2BtF1cY2wXAglsgqKrXW6ImJQhZQgI%2BZqcDWWSOJmipibTi49ln0JO6U6a1uoijCU-MQLimycPNfJV9RQJSi36gC8XwAMe-V4k1klA0oBQcigKDA%2BD4zkUSFEnleXFyxHyRiKJTamHESnOl2ut3olRWQQOTQ2LWx5IGQxuG1aQA)

```python
from typing import overload, Any, Sequence

def foo(axis: Any):
    match axis:
        case None:
            return None
        case []:
            return ()  # ❌ Unreachable
        case int(a):
            return (a,)
        case _:
            return tuple(axis)
            
def bar(axis: Any):
    if axis is None:
        return None
    if isinstance(axis, Sequence) and len(axis)==0:
        return ()  # ✅ not considered unreachable
    if isinstance(axis, int):
        return (axis,)
    return tuple(axis)
```

**Comment by sterliakov** [comment]
The original problem no longer reproduces on latest 1.14.1, but the next snippet is still broken and thinks that empty sequence is unreachable.
